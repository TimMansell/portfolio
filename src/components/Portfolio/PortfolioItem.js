import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';

import Button from 'components/Button';
import PortfolioIcons from './PortfolioIcons';
import Picture from '../Picture';
import Labels from '../Labels';

import useInViewport from 'hooks/useInViewport';

import styles from './PortfolioItem.module.scss';

export const PortfolioItem = ({
  src,
  title,
  year,
  description,
  tech,
  url,
  source,
}) => {
  const [cubeStyles, setCubeStyles] = useState({});
  const cubeElement = useRef(null);
  const isInViewport = useInViewport(cubeElement, 300);

  const { name, types, fallback } = src;
  const srcs = types.map((type) => ({
    type,
    src: require(`./img/${name}.${type}`),
  }));
  const defaultImg = require(`./img/${name}.${fallback}`);

  const getCubeHeight = debounce(() => {
    const { clientHeight } = cubeElement.current;

    setCubeStyles({
      transformOrigin: `50% 50% -${clientHeight / 2}px`,
    });
  });

  // Check height of image as intially it is lazy loaded.
  useEffect(() => {
    window.requestAnimationFrame(getCubeHeight);

    return () => window.cancelAnimationFrame(getCubeHeight);
  });

  return (
    <div className={styles.portfolioItem}>
      <div className={styles.cube} style={cubeStyles} ref={cubeElement}>
        <div className={styles.browser}>
          <PortfolioIcons title={title} />
          {isInViewport && (
            <Picture srcs={srcs} name={title} defaultImg={defaultImg} />
          )}
        </div>

        <div className={styles.info}>
          <h3 className={styles.title}>{title}</h3>
          <p className={styles.description}>{description}</p>
          <div className={styles.buttons}>
            <Button
              href={url}
              target="_blank"
              title="Visit website"
              data-test="portfolio-website-btn"
            >
              Visit website
            </Button>
            {source && (
              <Button
                href={source}
                target="_blank"
                title="View Souce"
                data-test="portfolio-source-btn"
              >
                View source
              </Button>
            )}
          </div>
        </div>
      </div>
      <div className={styles.tech} data-test="portfolio-tech">
        <Labels items={[year]} type="secondary" />
      </div>
      <Labels items={tech} type="primary" />
    </div>
  );
};

export default PortfolioItem;

PortfolioItem.propTypes = {
  src: PropTypes.object.isRequired,
  title: PropTypes.string.isRequired,
  year: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  tech: PropTypes.array.isRequired,
  url: PropTypes.string.isRequired,
  source: PropTypes.string,
};
