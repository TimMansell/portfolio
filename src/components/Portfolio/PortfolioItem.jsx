import { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';

import Button from 'components/Button';
import PortfolioIcons from './PortfolioIcons';
import Picture from '../Picture';
import Labels from '../Labels';

import styles from './PortfolioItem.module.scss';

export const PortfolioItem = ({
  img,
  title,
  years,
  types,
  description,
  tech,
  url,
  source,
}) => {
  const [cubeStyles, setCubeStyles] = useState({});
  const cubeElement = useRef(null);

  const getCubeHeight = debounce(() => {
    const { clientHeight } = cubeElement.current;

    setCubeStyles({
      transformOrigin: `50% 50% -${clientHeight / 2}px`,
    });
  });

  const labels = [
    {
      type: 'secondary',
      items: [...years, ...types],
    },
    {
      type: 'primary',
      items: tech,
    },
  ];

  // Check height of image as intially it is lazy loaded.
  useEffect(() => {
    window.requestAnimationFrame(getCubeHeight);

    return () => window.cancelAnimationFrame(getCubeHeight);
  });

  return (
    <div className={styles.portfolioItem}>
      <div className={styles.cube} style={cubeStyles} ref={cubeElement}>
        <div className={styles.browser}>
          <PortfolioIcons title={title} />
          <Picture
            src={`Portfolio/img/${img}`}
            title={`${title} portfolio item`}
            types={['webp', 'jpg']}
            srcSizes={[
              {
                sizes: ['480', '640', '768', '1024', '1280'],
              },
            ]}
            sizes="(min-width: 1200px) 33vw, (min-width: 768px) 50vw, 100vw"
            isLazy
          />
        </div>

        <div className={styles.info}>
          <h3 className={styles.title}>{title}</h3>
          <p className={styles.description}>{description}</p>
          <div className={styles.buttons}>
            <Button
              href={url}
              target="_blank"
              title="Visit website"
              data-test="portfolio-website-btn"
            >
              Visit website
            </Button>
            {source && (
              <Button
                href={source}
                target="_blank"
                title="View Souce"
                data-test="portfolio-source-btn"
              >
                View source
              </Button>
            )}
          </div>
        </div>
      </div>
      <div className={styles.labels}>
        <Labels labels={labels} />
      </div>
    </div>
  );
};

export default PortfolioItem;

PortfolioItem.propTypes = {
  img: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  years: PropTypes.array.isRequired,
  types: PropTypes.array.isRequired,
  description: PropTypes.string.isRequired,
  tech: PropTypes.array.isRequired,
  url: PropTypes.string.isRequired,
  source: PropTypes.string,
};
