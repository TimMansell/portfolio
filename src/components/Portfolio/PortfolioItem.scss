@import 'variables';
@import 'utilities/breakpoints';

.portfolio-item {
	background: rgba($primary-colour, .3);
	border: 1px solid rgba($primary-colour, .305);
	padding: 10px;
	transition: all .3s;
	position: relative;
	overflow: hidden;	
	height: 0;
	padding-bottom: 56.25%;

    &:hover &__info {
    	transform: translateY(0);
    }

	&__link {
  		&:before {
  			content: '';
				width: 100%;
				height: 100%;
				position: absolute;
				transition: all .3s;
  		}
	}

	&__icon {
	  // color: $icon-colour;
	  padding-left: 5px;
	  padding-right: 5px;
	  transition: all .3s;

	  &:hover {
	  	transform: scale(1.2);
	  }
	}

	&__info {
		position: absolute;
  		transform: translateY(100%);
  		text-shadow: -1px -1px darken($primary-colour, 10%);
  		transition: all .3s;
  		top: 0;
  		left: 0;
  		background: rgba($primary-colour, .95);
  		height: 100%;
  		padding: 15px;
  		width: 100%;
		  color: $secondary-text-colour;
		  text-align: center;
	}

	&__title {
		margin-bottom: 10px;
	}

	&__description {
		font-size: 15px;
		line-height: 1.2;
		margin-bottom: 5px;
		display: none;

  		@include media-breakpoint-up(sm){
			display: block;
  		}
	}

	&__tech {
		position: absolute;
		bottom: 20px;
		left: 0;
		width: 100%;
		padding-left: 10%;
		padding-right: 10%;
		font-size: 15px;
		line-height: 1.2;
	}

	&__more-info {
		position: absolute;
		bottom: 10px;
		width: 100%;
		left: 0;
	}
}