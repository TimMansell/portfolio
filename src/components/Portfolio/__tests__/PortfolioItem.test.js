import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import PortfolioItem from '../PortfolioItem';

const defaultProps = {
  title: 'Title',
  year: '2020',
  description: 'Description',
  tech: ['Tech', 'Tech 2'],
  url: 'url.com',
  src: {
    name: 'solitaire-project',
    types: ['avif', 'webp'],
    fallback: 'jpg',
  },
  source: 'source',
};

describe('PortfolioItem', () => {
  it('should render my component', () => {
    const props = {
      ...defaultProps,
    };
    // eslint-disable-next-line
    const wrapper = shallow(<PortfolioItem {...props} />);
  });

  it('should match snapshot', () => {
    const props = {
      ...defaultProps,
    };
    const snapshot = renderer.create(<PortfolioItem {...props} />).toJSON();

    expect(snapshot).toMatchSnapshot();
  });

  it('should render source button', () => {
    const props = {
      ...defaultProps,
    };
    const wrapper = shallow(<PortfolioItem {...props} />);
    const btn = wrapper.find('[data-test="portfolio-website-btn"]');

    expect(btn.exists()).toBeTruthy();
  });

  it('should not render source button', () => {
    const { title, year, description, tech, url, src } = defaultProps;

    const props = {
      title,
      year,
      description,
      tech,
      url,
      src,
    };
    const wrapper = shallow(<PortfolioItem {...props} />);
    const btn = wrapper.find('[data-test="portfolio-source-btn"]');

    expect(btn.exists()).toBeFalsy();
  });

  it('should render correct amount of tech labels', () => {
    const props = {
      ...defaultProps,
    };
    const wrapper = shallow(<PortfolioItem {...props} />);
    const tech = wrapper.find('[data-test="portfolio-tech"]');

    expect(tech.children()).toHaveLength(3);
  });
});
