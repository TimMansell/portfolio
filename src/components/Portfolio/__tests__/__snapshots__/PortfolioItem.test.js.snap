// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortfolioItem should match snapshot 1`] = `
<div
  className="portfolioItem"
>
  <div
    className="cube"
    style={Object {}}
  >
    <div
      className="browser"
    >
      <div
        className="bar"
      >
        <div
          className="icons"
        >
          <div
            className="icon"
          />
          <div
            className="icon"
          />
          <div
            className="icon"
          />
        </div>
        <div
          className="tab"
        >
          Title
        </div>
      </div>
      <div>
        <span>
          Â 
        </span>
      </div>
    </div>
    <div
      className="info"
    >
      <h3
        className="title"
      >
        Title
      </h3>
      <p
        className="description"
      >
        Description
      </p>
      <div
        className="buttons"
      >
        <a
          className="btn"
          href="url.com"
          rel="noopener noreferrer"
          target="_blank"
          title="Visit website"
        >
          Visit website
        </a>
        <a
          className="btn"
          href="source"
          rel="noopener noreferrer"
          target="_blank"
          title="View Souce"
        >
          View source
        </a>
      </div>
    </div>
  </div>
  <div
    className="tech"
    data-test="portfolio-tech"
  >
    <div
      className="label secondary"
    >
      2020
    </div>
    <div
      className="label primary"
    >
      Tech
    </div>
    <div
      className="label primary"
    >
      Tech 2
    </div>
  </div>
</div>
`;
